## Typescript: Complex Data Types, JSON, Map, Filter, Reduce

## PPL 2020 - http://www.cs.bgu.ac.il/~ppl202 
## Practical Session - Week #1

## How to Install Node.js and NPM on Windows

Node.js is a JavaScript interpreter. NPM (Node Package Manager) let's you install software (libraries, plugins, frameworks and applications) used to build Node applications. Both are installed using the installer package avaiable from the Node.js web site. The isntallation is pretty straightforward: 

1. Download the Windows installer from [nodejs.org](https://nodejs.org/en/)
2. Run the installer (the .msi file you downloaded in the previous step.)
3. Follow the prompts in the installer (Accept the license agreement, click the NEXT button a bunch of times and accept the default installation settings).
4. Restart your computer.

#### Test the installation: 
* Open the Windows Command Prompt (press the Win + R keys. Then, type cmd and press Enter).
* To verify node is installed, type "node -v", then press Enter. You should see a version number (e.g., v12.16.1 or v13.11.0).
* To verify NPM installation, type "npm -v" and press Enter. You should see a version number (e.g., 6.13.4 or 6.13.7).

#### More information:
For instructions on how to install other software that will be used throughout the course, please refer to the [useful links](https://www.cs.bgu.ac.il/~ppl202/Useful_Links) section in the course web site.  In particular, see instructions on how to install the TypeScript compiler tsc and the VSCode IDE.

## 1. Types of Values

A **data-type** is a classification of data which indicates what the programmer intends to do with the data. Types correspond to a set of values (eg. the boolean type corresponds to {true,false}, number type corresponds to the infinite set of numeric values), and they define the operations that can be done on the data (eg. booleans can be computed using logical operators).

We pay attention to two distinguishable defintions: The **type of a value** (atomic/compound) and the **type of a variable**. The type of a variable may be declared by the programmer of inferred by the type of the value it stores. The type of a value is determined by a defined group to which it belongs. The type of a variable may be different than the type of the value it stores, but they must be compatiable.

Apart from atomic types (number, string, boolean), there are **compound types**. Javascript introduces two basic compound types: Arrays and Maps. In Javascript, compound types are called *object* types (note: not to be confused with the more standard term - *object* here simply refers to any value which is not an atomic type)

##### Arrays

Arrays are the first basic compound type in Javascript. Here is an example of using arrays:

In [1]:
// This is how we declare an array containing the elements 16,8,27,13
let a = [16, 8, 27, 13]

// Here's how we use the Array compound value getter:
console.log(a[3])

// There is also a slice method:
console.log(a.slice(1,4))

13
[ 8, 27, 13 ]


undefined

Let's check that it is a compound type:

In [2]:
console.log(typeof(a))

object


undefined

And now the more specific reflection operator: 

In [3]:
console.log(a instanceof Array)

true


undefined

The array we defined above is **homogeneous**, meaning that it contains a single data type. 
The **homogeneous** array `a` above contains values whose types are `number`. 
Arrays in Javascript can be heterogeneous as well: 

In [4]:
let manyArray = [1, 'a', '2', "The Dawn", true]

console.log(`manyArray = ${manyArray}`)

manyArray = 1,a,2,The Dawn,true


undefined

##### Maps and the Object.keys

Maps are the second basic compound type in Javascript:

In [5]:
// This is how we declare them
let map = {a: 1, b:2}

// This is how we use the Map compound value getter for the values:
console.log(map['a'])
console.log(map.a)
console.log(map['b'])
console.log(map.b)

// And for keys (relevant for both Arrays [the keys there are the indexes] and Maps)
console.log(Object.keys(a))
console.log(Object.keys(map))

1
1
2
2
[ '0', '1', '2', '3' ]
[ 'a', 'b' ]


undefined

##### Variable Types

Javascript is a *dynamic language*, meaning that variables are not typed. But when a variable is bound to a value, we can inspect its type at runtime. 

Typescript extends Javascript to introduce optional variable types - it is compiled into Javascript and at compilation time, type checking is performed. Let's recall the example from the lectures: 

In [6]:
let typedVarNum  : number = 6,
    typedVarStr  : string = "blue",
    typedVarBool : boolean = true;

undefined

## 2. JSON

### 2.1 JSON Review

JSON stands for Java Script Object Notation and is a standard way to serialize Javascript compound values into strings (and vice-versa). It is widely used to allow data exchange in server-client communication. It serves a role that is similar to XML.

JSON is a **notation** - that is, a way of writing Javascript values in a way that allows reading them back easily. 
Most Javascript values can be written in JSON - but not all of them.

Once you have a Javascript object, and once there is an agreed-upon-way of "stringifying" the object (in our case, the standard way is `JSON.stringify()`), then you can easily pass this string through the network, from a computer to a computer, as passing strings through the network is easy, and the receiving side can only do an agreed-upon way to parse the string into an object (in our case, the standard way is `JSON.parse()`).

The JSON interface defines the following two methods: 
1. `JSON.stringify(o)` - maps a value to a string
2. `JSON.parse(s)` - maps a string written according to the JSON syntax to a value

JSON serves a similar goal to that of `Serializable` of Java that we saw in SPL.

In [7]:
let person1 = { name : "Yosi", age : 31, city : "Beer Sheva" },
    person1JSONString = JSON.stringify(person1);

console.log(`person1 serialized in JSON = ${person1JSONString}`)
console.log(`person1JSONString is of type ${typeof person1JSONString}`)

person1 serialized in JSON = {"name":"Yosi","age":31,"city":"Beer Sheva"}
person1JSONString is of type string


undefined

In [8]:
let person2 = JSON.parse(person1JSONString)

console.log(person2)
console.log(`person2 is of type ${typeof person2}`)

{ name: 'Yosi', age: 31, city: 'Beer Sheva' }
person2 is of type object


undefined

Note that the way values are written in JSON is not exactly the same as we write them in a Javascript program - there are quotes ("") around keys for example.

Note also that `stringify()` and `parse()` work on atomic values as well:

In [9]:
console.log(typeof(JSON.parse('2')))     // note that the input to JSON.parse is a JSON string
console.log(typeof(JSON.parse('true')))
console.log(typeof(JSON.parse('"abc"')))

number
boolean
string


undefined

The values that can be encoded into JSON (let us call them "JSON values") must be one of the following types:
1. string
2. number
3. boolean
4. null
5. a map where all the keys are strings and all the values are JSON values (recursively) 
6. an array where all the values are JSON values

The Javascript compound data types, *arrays* and *maps*, can be combined in a recursive manner. 
This applies to JSON as well.

The Javascript values that **cannot be encoded** into JSON strings are `undefined` and functions.

Let us look at an example of a complex value and how we can describe it in JSON and then how using types can help us describe its structure in a well-documented manner:

In [10]:
let studentsData = {department: "Computer Science", 
                    students: [
                        {name: "James", degree:"MSc"}, 
                        {name: "John", degree:"PhD"}
                    ]};
let studentsJSON = JSON.stringify(studentsData)
console.log(studentsJSON)
console.log(`studentsJSON is of type ${typeof studentsJSON}`)
let studentsJSONParsed = JSON.parse(studentsJSON)
console.log(studentsJSONParsed)
console.log(`studentsJSONParsed is of type ${typeof studentsJSONParsed}`)

{"department":"Computer Science","students":[{"name":"James","degree":"MSc"},{"name":"John","degree":"PhD"}]}
studentsJSON is of type string
{ department: 'Computer Science',
  students: 
   [ { name: 'James', degree: 'MSc' },
     { name: 'John', degree: 'PhD' } ] }
studentsJSONParsed is of type object


undefined

We verified that this complex structure can be written into JSON, and then parsed back into an identical value (this is a **round-trip** that preserves the value).

We can look at the type of studentData as a **tree** of simpler values:
* `studentsData` is a map which has 2 keys:
  * `department` with a value which is a string
  * `students` with a value which is an embedded array whose items are compound values.
* The nested values are maps that have 2 keys:
  * `name` whose value is a `string` 
  *  `degree` whose value is also a `string`
  

### 2.2 Documenting Complex Values with Type Annotations

We can describe this structure using the following TypeScript declaration:
```
{ department:string;
  students: {name:string, degree:string}[];
}
```

This can be made more readable if we name the types, using the **interface** TypeScript construct to name map compound types:

In [11]:
interface Student {
    name: string;
    degree:string;
};

interface StudentsData {
    department: string;
    students: Student[];
}

undefined

This type definition is useful to document the type of the expected values that we expect to read.

### 2.3 Working on complex JSON values

In this section, we will use the methods **map,filter,reduce** to operate over complex JSON values. 

We will take a videos-database example (http://reactivex.io/learnrx/) as a basis for our examples.

Suppose you are a developer at the popular streaming-movie website Netflix, and that your system's *API* uses JSON to communicate data.  The JSON values we describe are actual JSON values returned by the Netflix API.

When searching for new movie releases, you send a query to the API, and obtain a JSON reply:

In [12]:
let newReleases = [
        {
            "id": 70111470,
            "title": "Die Hard",
            "boxart": "http://cdn-0.nflximg.com/images/2891/DieHard.jpg",
            "uri": "http://api.netflix.com/catalog/titles/movies/70111470",
            "rating": 4.0,
            "bookmark": []
        },
        {
            "id": 654356453,
            "title": "Bad Boys",
            "boxart": "http://cdn-0.nflximg.com/images/2891/BadBoys.jpg",
            "uri": "http://api.netflix.com/catalog/titles/movies/70111470",
            "rating": 5.0,
            "bookmark": [{ id: 432534, time: 65876586 }]
        },
        {
            "id": 65432445,
            "title": "The Chamber",
            "boxart": "http://cdn-0.nflximg.com/images/2891/TheChamber.jpg",
            "uri": "http://api.netflix.com/catalog/titles/movies/70111470",
            "rating": 4.0,
            "bookmark": []
        },
        {
            "id": 675465,
            "title": "Fracture",
            "boxart": "http://cdn-0.nflximg.com/images/2891/Fracture.jpg",
            "uri": "http://api.netflix.com/catalog/titles/movies/70111470",
            "rating": 5.0,
            "bookmark": [{ id: 432534, time: 65876586 }]
        }
    ];

undefined

Let us describe the **type** of this value:
* We observe it is a homogeneous array of map values.
* Each map value describes a "movie" with fields id, title, boxart, uri, rating and bookmark.

The corresponding Typescript annotation is:

In [13]:
interface Video {
    id: number;
    title: string;
    boxart: string;
    uri: string;
    rating: number;
    bookmark: {id: number, time: number}[];
}
type Releases = Video[];

undefined

This type annotation lets us think about the data we obtained from the API as an array `[v1, v2, v3, v4]` where each item $v_i$ is a map `{id, title, boxart, uri, rating, bookmark}`.


Note: while *interface* actually introduces a new type, *type* only provides an alias and not a new type. For instance, alias names will not be used in error messages. In addition, unlike an interface, one can give a type alias to any type annotation.

#### 2.3.1 Map over an Array of Videos

We would like to transform this data into an array of `{id,title}` properties. 
This transformation is called a **projection** as we select only some of the properties in the input.

In terms of types, this transformation maps from the type `Video[]` into the type `{id:number, title:string}[]`.

A procedural way to achieve this goal would be:

In [14]:
function getIDsAndTitles_1(reply)
{
    let res = [];
    for(let i=0; i<reply.length; i++)
    {
        res.push({id: reply[i].id, title: reply[i].title});
    }
    return res;
}

let newReleasesIDAndTitle = getIDsAndTitles_1(newReleases)
console.log(newReleasesIDAndTitle)

[ { id: 70111470, title: 'Die Hard' },
  { id: 654356453, title: 'Bad Boys' },
  { id: 65432445, title: 'The Chamber' },
  { id: 675465, title: 'Fracture' } ]


undefined

A functional way of achieving the same result uses the **map** method (seen in the reading material for week-1) to abstract away the loop and the mutations we observe in the procedural solution (`i++, res.push()`).

In [15]:
let newReleasesIDAndTitle = newReleases.map(x => ({id: x.id, title: x.title}))
console.log(newReleasesIDAndTitle)

[ { id: 70111470, title: 'Die Hard' },
  { id: 654356453, title: 'Bad Boys' },
  { id: 65432445, title: 'The Chamber' },
  { id: 675465, title: 'Fracture' } ]


undefined

Note that in this version:
* There is no loop (`map` abstracts away the loop control structure)
* There is no variable assignment (no need to define a loop counter `i` and mutate it and to maintain a `res` accumulator variable).

Let us note what is the type of the transformer function we pass as an argument to `map()`:
it gets as an argument `x` an item from newReleases - which is of type `Video` - and it returns a map of type `{id:number, name:title}`.  

#### 2.3.2 Filter an Array of Videos

The user is quite picky, and would like to search for new releases that have a rating of 5.0 only. 

We can solve this problem using **filter**.

In terms of types, the operation we want to define maps an array of `Video[]` to an array of `Video[]`.
This fits the definition of `filter` - which does not change the type of items in the array.

In [16]:
let newReleasesOfRating5 = newReleases.filter(x => x.rating === 5)
console.log(newReleasesOfRating5)

[ { id: 654356453,
    title: 'Bad Boys',
    boxart: 'http://cdn-0.nflximg.com/images/2891/BadBoys.jpg',
    uri: 'http://api.netflix.com/catalog/titles/movies/70111470',
    rating: 5,
    bookmark: [ [Object] ] },
  { id: 675465,
    title: 'Fracture',
    boxart: 'http://cdn-0.nflximg.com/images/2891/Fracture.jpg',
    uri: 'http://api.netflix.com/catalog/titles/movies/70111470',
    rating: 5,
    bookmark: [ [Object] ] } ]


undefined

#### 2.3.4 Reminder: Reduce and Neutral Elements as Initializers
Compare the following 5 invocations of `reduce`:

In [1]:
// Compute the sum of an array of integers
[1, 2, 3].reduce((acc, cur) => acc + cur, 0)

6

In [1]:
// Compute the product of an array of integers
[1, 2, 3].reduce((acc, cur) => acc * cur, 1)

6

In [1]:
// Compute the logical and of an array of booleans
[true, false, true].reduce((acc, cur) => acc && cur, true)

false

In [1]:
// Compute the logical or of an array of booleans
[true, false, true].reduce((acc, cur) => acc || cur, false)

true

In [1]:
// Compute the max of an array of numbers
[3, 1, 4].reduce((acc,cur) => Math.max(acc,cur), 0)

4

Note that in all 5 cases, the initializer passed to `reduce(reducer,init)` is the neutral element of the reducer operator
(0 for `+`, 1 for `*`, `true` for `and`, `false` for `or`).

This makes sense in general - and allows one to answer the puzzling question of **what should be the value of reduce when applied to an empty array**.  The general answer is that it should be the neutral element of the reducer.

**NOTE**: the `max` version takes as an initializer the value 0.  This is not correct when the array can contain negative numbers.  What should be the initial value in this case?

#### 2.3.5 Rectangle Selection with Reduce

A user has chosen to view movies as boxarts (that is, the image that is shown for each movie), as it is easier to select movies according to their boxarts. The API returned the following value as a reply to our query:

In [1]:
let boxarts = [
        { width: 200, height: 200, url: "http://cdn-0.nflximg.com/images/2891/Fracture200.jpg" },
        { width: 150, height: 200, url: "http://cdn-0.nflximg.com/images/2891/Fracture150.jpg" },
        { width: 300, height: 200, url: "http://cdn-0.nflximg.com/images/2891/Fracture300.jpg" },
        { width: 425, height: 150, url: "http://cdn-0.nflximg.com/images/2891/Fracture425.jpg" }
    ]

undefined

Let us define the type of this returned value: it is a homogeneous array of maps.
Each map describes a boxart.

```
interface Boxart {
    width:number; 
    height:number;
    url: string;
}
```

We would like to find the largest box-art image size, so that we could know what is the maximal size of the image placeholder should be.  We measure rectangles by their area ($width \times height$).

A straightforward way to do this is by going over each boxart, and keeping a temporary variable that holds the largest size out of the boxarts we have seen so far.

In terms of types, the transformation maps an array `Boxart[]` into a `number`. Let us annotate this type into the function definition:

In [2]:
interface Boxart {
    width:number; 
    height:number;
    url: string;
};

function maxBox(boxes: Boxart[]):number {
    let maxBox = 0;
    for (let i = 0; i < boxes.length; i++) {
        let curBoxSize = boxes[i].width * boxes[i].height;
        if (curBoxSize > maxBox) {
            maxBox = curBoxSize;
        }
    }
    return maxBox;
}
maxBox(boxarts)

63750

This is a case of a loop to accumulate a value from an array of values. In terms of types, it maps an array to a single value. The **reduce** method fits perfectly for this scenario: The `reduce` method takes as parameter a reducer function and an initial value, and returns the accumulated value. 

The reducer function takes two arguments: the current accumulator, and the current item. The return value of the reducer function is the "successive" value of the accumulator parameter. That is, the result of some operation applied to the current accumulator, and the current item.

Our current accumulator should be the current largest box-art image size, and the initial value should be 0.

In [3]:
let largestBoxartSize = boxarts.reduce(
    (curMax,curBox) => {
        let curBoxSize = curBox.width * curBox.height;
           if (curBoxSize > curMax) {
               return curBoxSize;
           } else {
               return curMax;
           }
    }, // this is the reducer function
    0  // this is the initial value
    );
    
console.log(largestBoxartSize)

63750


undefined

Indeed, the largest boxart image size is 63750, which is the 4th boxart.

Note that this version has no side effect, no mutation and no loop.

We can make this version more readable by using the Math.max function - to express clearly that what we are doing in this 
`reduce` invocation is the identification of the max of an array.  We also annotate the expected type of the parameters in the reducer.

In [25]:
const boxSize = (box:Boxart) => box.width * box.height;
let maxBoxSize = boxarts.reduce((curMax:number, curBox:Boxart) => Math.max(curMax, boxSize(curBox)), 0);
   
console.log(maxBoxSize)

63750


undefined

### 2.4 Tree Values

Some replies will be more complex than what we have witnessed above: it is very often that we face replies that have a **tree** form, as opposed to a flat homogeneous array.

One example of such, is the following API reply - which returns values grouped by movie genre ("New Releases", "Dramas" ...)

In [26]:
let movieLists = [
        {
            name: "New Releases",
            videos: [
                {
                    "id": 70111470,
                    "title": "Die Hard",
                    "boxart": "http://cdn-0.nflximg.com/images/2891/DieHard.jpg",
                    "uri": "http://api.netflix.com/catalog/titles/movies/70111470",
                    "rating": 4.0,
                    "bookmark": []
                },
                {
                    "id": 654356453,
                    "title": "Bad Boys",
                    "boxart": "http://cdn-0.nflximg.com/images/2891/BadBoys.jpg",
                    "uri": "http://api.netflix.com/catalog/titles/movies/70111470",
                    "rating": 5.0,
                    "bookmark": [{ id: 432534, time: 65876586 }]
                }
            ]
        },
        {
            name: "Dramas",
            videos: [
                {
                    "id": 65432445,
                    "title": "The Chamber",
                    "boxart": "http://cdn-0.nflximg.com/images/2891/TheChamber.jpg",
                    "uri": "http://api.netflix.com/catalog/titles/movies/70111470",
                    "rating": 4.0,
                    "bookmark": []
                },
                {
                    "id": 675465,
                    "title": "Fracture",
                    "boxart": "http://cdn-0.nflximg.com/images/2891/Fracture.jpg",
                    "uri": "http://api.netflix.com/catalog/titles/movies/70111470",
                    "rating": 5.0,
                    "bookmark": [{ id: 432534, time: 65876586 }]
                }
            ]
        }
    ]

undefined

To document the structure of this value, let us write its type:
* It is an array of 2 maps of identical structure.
* Each item represents a category of Videos. The category maps have the following keys:
  * name: string 
  * videos: an array of Video[].

The Typescript definition is thus:

In [27]:
interface VideoCategory {
    name:string;
    videos: Video[];
}
type VideoCategories = VideoCategory[];

undefined

The structure of the value is thus a *tree*:
$[ category_1, category_2 ]$
where 
$category_i$ is of the form:
`{name:string, videos:[v1,...]}`.

These kinds of trees impose a challenge when working with them: if we apply `map` or `filter`, the transformer or the predicate will receive as argument a `VideoCategory` object.  If we want to process the embedded `Video` objects inside the categories, we must first **flatten** the tree.

Suppose we would like to have a list of all movie ids in that tree. The most trivial way, as before, is to do the following (using a foreach loop):

In [39]:
let movieIds = []
movieLists.forEach(category => category.videos.forEach(video => movieIds.push(video.id)));
console.log(`movieIds=${movieIds}`);

movieIds=70111470,654356453,65432445,675465


undefined

`forEach()` abstracts a single loop.  The embedded calls to `forEach` inside `forEach` reflects the two-level traversal of the tree.

Note that `forEach()` is **not a functional** style: this is precisely the type of operations we want to abstract away using map/filter/reduce. This version has an accumulator variable (`movieIds`) which we must initialize and then mutate as part of the `forEach` iteration.

Not a functional solution...

The general feeling of the task we want to perform is an **accumulation** - so that `reduce` seems to be the good tool.
But we cannot apply `reduce()` as is, because `reduce` would work on `Category` values and we want to accumulate `Video` values.

Alternatively, we could attempt to use **map** as we are mapping `Video` values into `number` values.  But again, we cannot map over Categories - we need to map over Videos - so that `map` cannot be used directly.

Let's recall the concat() method of arrays:

In [40]:
let arrayOne = [1, 2, 3, 4]
let arrayTwo = [5, 6, 7, 8]
arrayOne.concat(arrayTwo)

[ 1, 2, 3, 4, 5, 6, 7, 8 ]

In contrast to `push()`, `concat()` does not mutate its arguments - it returns a new array and can be used safely.

If we `reduce` an array of arrays and accumulate using the `concat` operator,  we will get a method that flattens one level of an array of arrays:

In [41]:
let a = [[1, 2, 3], [4, 5, 6]]
a.reduce((acc, curr) => acc.concat(curr), [])

[ 1, 2, 3, 4, 5, 6 ]

We now need to apply a mixture of `map` and `reduce` in order to map and flatten:

In [42]:
// We want the following as an array of the numbers:
let a = [{group: 1, numbers:[1, 2, 3]}, {group: 2, numbers:[4, 5, 6]}];
a.map(x => x.numbers).reduce((acc,curr) => acc.concat(curr), [])

[ 1, 2, 3, 4, 5, 6 ]

### 2.4.1 The 'ramda' package

'ramda' is a library of functions designed tofacilitate functional programming in JavaScript.

ramda includes a function called `chain()` (often also called `flatmap()`) which can come handy in our last example: If we want to return all of the results as a single flat array instead of as an array of arrays, then we can use `ramda`'s chain:

In [44]:
import * as R from 'ramda'

undefined

In [45]:
R.chain(x => x.numbers, a)

[ 1, 2, 3, 4, 5, 6 ]

The `chain()` function goes through an array, and flattens the result of 'doing something' (which can be 'nothing' or applying a 'mapping') by one level.

Back to our videos example, if we `chain` the `movieLists` with a function that just returns the category videos, then the result would be an array containing the videos. As such:

In [46]:
R.chain(category => category.videos, movieLists)

[ { id: 70111470,
    title: 'Die Hard',
    boxart: 'http://cdn-0.nflximg.com/images/2891/DieHard.jpg',
    uri: 'http://api.netflix.com/catalog/titles/movies/70111470',
    rating: 4,
    bookmark: [] },
  { id: 654356453,
    title: 'Bad Boys',
    boxart: 'http://cdn-0.nflximg.com/images/2891/BadBoys.jpg',
    uri: 'http://api.netflix.com/catalog/titles/movies/70111470',
    rating: 5,
    bookmark: [ [Object] ] },
  { id: 65432445,
    title: 'The Chamber',
    boxart: 'http://cdn-0.nflximg.com/images/2891/TheChamber.jpg',
    uri: 'http://api.netflix.com/catalog/titles/movies/70111470',
    rating: 4,
    bookmark: [] },
  { id: 675465,
    title: 'Fracture',
    boxart: 'http://cdn-0.nflximg.com/images/2891/Fracture.jpg',
    uri: 'http://api.netflix.com/catalog/titles/movies/70111470',
    rating: 5,
    bookmark: [ [Object] ] } ]

The only change that is left, is that we do not want the category videos flattened, rather something extra: taking the category.videos and mapping each element there to its id.

In [47]:
R.chain(category => category.videos, movieLists).map(video => video.id) 

[ 70111470, 654356453, 65432445, 675465 ]

It helps to annotate the types of the arguments in such chained transformations to document and verify their correctness:

In [48]:
R.chain((category:VideoCategory) => category.videos, movieLists).map((video:Video) => video.id) 

[ 70111470, 654356453, 65432445, 675465 ]

## 3. Exercises: 

### 3.1 Implement the `map` method using `reduce`.

Let's think step by step: What should be our *accumulated* value? It is natural to think that the accumulation consists in 'pushing' the current element **as a mapped value** to the currently accumulated array.

To that end, the initial value should be the empty array. As we go, we will take the current element, and our next accumulator will be the current accumulator, only expanded by the current element after it is mapped by the transformer.

[solution](PPL_PS1_SOL.html#Exercise-1:-Implementing-map-with-reduce)

### 3.2 Implement `filter` using `reduce`.

What should be changed in the solution of the previous exercise?

We can skip adding the current element if it does not satisfy the predicate, in which case the next accumulator will not change. However, if the current element does satisfy the predicate, we will set the next accumulator to be the current one, only enlarged by the current element

[solution](PPL_PS1_SOL.html#Exercise-2:-Implementing-filter-with-reduce)

### 3.3 Implement `some` and `every` using `map` and `reduce`.

`some(pred)` is a method that returns `true` if and only if *at least one* element in an array satisfies a given predicate.

`every(pred)` is a method that returns true if and only if *all* the elements in an array satisfy a given predicate.

For example:

In [49]:
let even = (x => x % 2 == 0)
let arr1 = [1,2,3,4]
let arr2 = [1,3,5,7]
let arr3 = [2,4,6,8]
let arr1HasEvenNumbers = arr1.some(even)
let arr2HasEvenNumbers = arr2.some(even)
let allInArr1AreEven = arr1.every(even)
let allInArr3AreEven = arr3.every(even)
console.log(`arr1HasEvenNumbers = ${arr1HasEvenNumbers}`)
console.log(`arr2HasEvenNumbers = ${arr2HasEvenNumbers}`)
console.log(`allInArr1AreEven = ${allInArr1AreEven}`)
console.log(`allInArr3AreEven = ${allInArr3AreEven}`)

arr1HasEvenNumbers = true
arr2HasEvenNumbers = false
allInArr1AreEven = false
allInArr3AreEven = true


undefined

Instructions: 

A way to implement `some` is by mapping the array to `true/false` values of whether each element satisfies the predicate, and then doing an accumulated logical `or` between the elements using `reduce`, with the initial accumulator equal to `false`.

Similarly, we can implement `every` in the same way as above, except that we need to use a logical `and` and our initial accumulator should be `true`.

[solution](PPL_PS1_SOL.html#Exercise-3:-Implementing-some-and-every-with-map-and-reduce)

## Shortcut Semantics

The native `some` and `every` methods employ a concept known as 'shortcut semantics'. What this means, is that `some` stops and immediately returns `true` at the moment it finds an element that satisfies the predicate. `every` stops and immediately returns `false` at the moment it finds an element that does not satisfy the predicate.

**Question**: does our implementation satisfy shortcut semantics?

It is difficult to distinguish between shortcut semantics and non-shortcut semantics because both shortcut and non-shortcut versions return the same values for all parameters.  

Can we conclude that a shortcut and non-shortcut versions of `some` are equivalent according to the definition of function equivalence we provided in class?

Remember though the difference between mathematical function equivalence and programming functions equivalence: 2 programming functions `f` and `g` are equivalent if they have the same domain, same range and for all values in the domain:
* Either `f(x)` has a value and `g(x)` has the same value
* Or `f(x)` does not terminate and `g(x)` does not terminate as well
* Or `f(x)` throws an error and `g(x)` throws an error as well

How can we then distinguish between a shortcut semantics version of `some` and one that is not shortcut?

The solution is to choose parameters that will distinguish between the two computations using either non-termination (infinite loops) or errors.

Let us define such an example on arrays of numbers.  We can trigger an error in Javascript by using the `throw` primitive.

To test whether an error was thrown, we use the same syntax as in Java - with `try/catch`.

In [50]:
const throwOnZero = x => { if (x > 0) return true; else throw false;}
let a = [1,0]
try {
    a.some(throwOnZero);
}
catch (e) {
    e;
}

true

In [51]:
const someExercise = (pred, arr) => arr.map(pred).reduce((acc, cur) => acc || cur, false)
try {
    someExercise(throwOnZero, a);
}
catch (e) {
    e;
}

false

By using deliberate error throwing, we are able to distinguish between shortcut and non-shortcut semantics on the definition 
of the `some` function.

The next question that arises is whether we can define a version of `some` using `reduce` that has shortcut semantics.
We will revisit this question in Chapter 3 of the course, when we introduce generators.